src/backend/cpu/convolve.cpp

```diff
@@ -107,5 +119,197 @@ INSTANTIATE(ushort ,   float)
+
+    Array<accT> collapsedFilter = filter;
+
+    vector<af_seq> flip_index(4);
+    af_seq s = {(double)(fDims[0] - 1), 0, -1};
+    flip_index[0] = s;
+    s = {(double)(fDims[1] - 1), 0, -1};
+    flip_index[1] = s;
+    flip_index[2] = af_span;
+    flip_index[3] = af_span;
+
+    collapsedFilter = createSubArray(collapsedFilter, flip_index);
+    collapsedFilter.modDims(dim4(fDims[0] * fDims[1] * fDims[2], fDims[3]));
+
+    Array<accT> res = matmul(collapsedFilter, unwrapped, AF_MAT_TRANS, AF_MAT_NONE);
+    res.modDims(dim4(collapsedFilter.dims()[1], out.dims()[0], out.dims()[1], signal.dims()[3]));
+    out = cast<T>(reorder(res, dim4(1, 2, 0, 3)));
+
+    return out;
+}
+
+
+template<typename T, typename accT>
+Array<T> convolve2(Array<T> const& signal, Array<accT> const& filter,
+                   const dim4 stride, const dim4 padding, const dim4 dilation) {
+    signal.eval();
+    filter.eval();
+
+    Array<T> out  = createEmptyArray<T>(dim4());
+    out = convolve2_unwrap<T, accT>(signal, filter, stride, padding, dilation);
+
+    return out;
+}
+
+#define INSTANTIATE(T, accT)                                                                \
+    template Array<T> convolve2<T, accT>(Array<T> const& signal, Array<accT> const& filter, \
+                                         const dim4 stride, const dim4 padding, const dim4 dilation);
+
+INSTANTIATE(cdouble, cdouble)
+INSTANTIATE(cfloat ,  cfloat)
+INSTANTIATE(double ,  double)
+INSTANTIATE(float  ,   float)
+INSTANTIATE(uint   ,   float)
+INSTANTIATE(int    ,   float)
+INSTANTIATE(uchar  ,   float)
+INSTANTIATE(char   ,   float)
+INSTANTIATE(ushort ,   float)
+INSTANTIATE(short  ,   float)
+INSTANTIATE(uintl  ,   float)
+INSTANTIATE(intl   ,   float)
+#undef INSTANTIATE
+
+template<typename T, typename accT>
+Array<T> conv2DataGradient(const Array<T>& incoming_gradient,
+                           const Array<T>& original_signal,
+                           const Array<accT>& original_filter,
+                           const Array<T>& convolved_output,
+                           af::dim4 stride, af::dim4 padding, af::dim4 dilation) {
+    original_signal.eval();
+    original_filter.eval();
+
+    const dim4 cDims = incoming_gradient.dims();
+    const dim4 sDims = original_signal.dims();
+    const dim4 fDims = original_filter.dims();
+
+    Array<accT> collapsed_filter = original_filter;
+
+    vector<af_seq> flip_index(4);
+    af_seq s = {(double)(fDims[0] - 1), 0, -1};
+    flip_index[0] = s;
+    s = {(double)(fDims[1] - 1), 0, -1};
+    flip_index[1] = s;
+    flip_index[2] = af_span;
+    flip_index[3] = af_span;
+
+    collapsed_filter = createSubArray(collapsed_filter, flip_index);
+    collapsed_filter.modDims(dim4(fDims[0] * fDims[1] * fDims[2], fDims[3]));
+
+    Array<accT> collapsed_gradient = cast<accT>(incoming_gradient);
+    collapsed_gradient = reorder(collapsed_gradient, dim4(0, 1, 3, 2));
```

Comment:
I don't understand the reason for this reorder. you are only changing the 3rd and 4th dims. I feel that we can do something more efficient here and in the next operations. maybe request the right shape of these functions.